EIPS/eip-non_wallet_keys.md

```diff
@@ -0,0 +1,52 @@
+---
+eip: <to be assigned>
+title: Non-wallet usage of keys derived from BIP-32 trees
+author: Michele Balistreri (@bitgamma)
+discussions-to: https://ethereum-magicians.org/t/non-wallet-usage-of-keys-derived-from-bip-32-trees/1817
+status: Draft
+type: Standards Track
+category: ERC
+created: 2018-11-13
+---
+
+## Simple Summary
+This EIP describes a derivation path structure for [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) wallets to be used for non-wallet keypairs.
+
+## Abstract
+BIP32 defines a way to generate hierarchical trees of keys which can be derived from a common master key. BIP32 and [BIP44](https://https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki) defines the usage of these keys as wallets. In this EIP we describe the usage of such keys outside the scope of the blockchain defining a logical tree for key usage which can cohesist (and thus share the same master) with existing BIP44 compatible wallets.
+
+## Motivation
+Applications interacting with the blockchain often make use of additional, non-blockchain technologies to perform the task they are designed for. For privacy and security sensitive mechanisms, sets of keys are needed. Reusing keys used for wallets can prove to be insecure, while keeping completely independent keys make backup and migration of the full set of credentials more complex. Defining a separate (from BIP44 compliant wallets) derivation branch allows combining the security of independent keys with the convenience of having a single piece of information which needs to be backup or migrated.
+
+## Specification
+
+### Path levels
+We define the following 3 levels in BIP32 path:
+
+```m / purpose' / key_type' / key_index```
+
+Apostrophe in the path indicates that BIP32 hardened derivation is used.
+
+Each level has a special meaning, described in the chapters below.
+
+### Purpose
+Purpose is a constant set to XXXX' (or 0x8000XXXX) following the [BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki) recommendation. It indicates that the subtree of this node is used according to this specification. XXXX is to be replaced by the EIP/ERC number.
+
+Hardened derivation is used at this level.
+
+
+### Key type
+Describes the purpose for which the key is being used. Key types should be generic. "Instant messaging" is a good example whereas "Whisper" is not. The reason is that you want to be able to use the same identity across different services. Key types are defined at: <to be defined>
```

Comment:
It is not clear to me why you would make a recommendation on keeping the same identity for all services. Whisper is designed so that you could generate a new identity per message, which is at odds with this approach. I understand that the key index is meant to do this, but why not leave the user decide what they want to do with each level?

I would take this comment even further and suggest that you let the user create as many stages in this process as they like. As such, I feel that it is too easy to just guess all the steps. The initial constant is easy to guess, the "Instant messaging" part too, and then the index is incremental. If you have more complex pipes with:

0x80001581 > "Instant messaging" > "Whisper" > "Interactions with Jenny" > "Music-related" > "name that band" > "Key Index", it's more difficult to predict.

And even then this pipeline is rather predictable in some cases, which could be improved by adding a random word at a random location in the pipe:

0x80001581 > "Instant messaging" > "Whisper" > "Interactions with Jenny" > "Music-related" > "Daffodil" > "name that band" > "Key Index"

Reply:
The idea is to give the option to reuse the same identity across different protocols if desired or use a different index if a new identity is preferable. For this reason I would have they key type categories to be generic ones.

The reason of having a relatively shallow hierarchy and predictable/fixed key path is for interoperability between client apps, making easy switching from one client to the other without losing contacts or other services associated to the identity at that specific path.

Could you please elaborate the treath of using predictable key paths? Knowing the key path alone does not mean knowing the identity since you can only derive the correct keys if you have the private key of the level above. Well-known paths are being used for wallets too. Has there been a change of direction on this lately?

If this is a concern, the EIP could be amended with your suggestion, so the first two levels are well-defined here, but the incremental index is replaced by a user-defined subtree. Then the user (and/or the client dev) can decide how to further structure keys, allowing both complex structures or simple indexes.